File: protocol/csharp/ToInfinity.Protocol/InputEvent.cs
using System.Buffers.Binary;

namespace ToInfinity.Protocol;

/// <summary>Mouse button identifiers for <see cref="MouseButtonDownEvent"/>/<see cref="MouseButtonUpEvent"/> (SPEC.md §4.2).</summary>
public enum MouseButton : byte
{
    Left = 0,
    Right = 1,
    Middle = 2,
    X1 = 3,
    X2 = 4,
}

/// <summary>SPEC.md §4.2 input event kind tags.</summary>
public enum InputEventKind : byte
{
    MouseMove = 0x01,
    MouseButtonDown = 0x02,
    MouseButtonUp = 0x03,
    MouseWheel = 0x04,
    KeyDown = 0x05,
    KeyUp = 0x06,
}

/// <summary>
/// Base type for binary input events carried on the control channel with
/// <see cref="ControlFrameKind.InputEvent"/> (SPEC.md §4). Mouse coordinates
/// are normalized 0-65535 per axis (SPEC.md §4.1); key codes are USB HID
/// Usage IDs, Usage Page 0x07 (SPEC.md §4.2).
/// </summary>
public abstract class InputEvent
{
    public abstract InputEventKind EventKind { get; }

    /// <summary>Encodes this event's payload (starting with the 1-byte event kind tag).</summary>
    public abstract byte[] Encode();

    /// <summary>
    /// Attempts to decode an input event payload (the bytes after the
    /// <see cref="ControlFrameKind.InputEvent"/> tag). Returns null (never
    /// throws) for an unrecognized event kind or a payload too short for
    /// its kind, per SPEC.md §6.
    /// </summary>
    public static InputEvent? TryDecode(ReadOnlySpan<byte> payload)
    {
        if (payload.Length < 1)
        {
            return null;
        }

        var kind = (InputEventKind)payload[0];
        switch (kind)
        {
            case InputEventKind.MouseMove:
                if (payload.Length != 5) return null;
                return new MouseMoveEvent(
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(1, 2)),
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(3, 2)));

            case InputEventKind.MouseButtonDown:
                if (payload.Length != 6) return null;
                return new MouseButtonDownEvent(
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(1, 2)),
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(3, 2)),
                    (MouseButton)payload[5]);

            case InputEventKind.MouseButtonUp:
                if (payload.Length != 6) return null;
                return new MouseButtonUpEvent(
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(1, 2)),
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(3, 2)),
                    (MouseButton)payload[5]);

            case InputEventKind.MouseWheel:
                if (payload.Length != 9) return null;
                return new MouseWheelEvent(
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(1, 2)),
                    BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(3, 2)),
                    BinaryPrimitives.ReadInt16BigEndian(payload.Slice(5, 2)),
                    BinaryPrimitives.ReadInt16BigEndian(payload.Slice(7, 2)));

            case InputEventKind.KeyDown:
                if (payload.Length != 3) return null;
                return new KeyDownEvent(BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(1, 2)));

            case InputEventKind.KeyUp:
                if (payload.Length != 3) return null;
                return new KeyUpEvent(BinaryPrimitives.ReadUInt16BigEndian(payload.Slice(1, 2)));

            default:
                return null;
        }
    }
}

/// <summary>SPEC.md §4.2 0x01 Mouse move (5 bytes).</summary>
public sealed class MouseMoveEvent : InputEvent
{
    public MouseMoveEvent(ushort x, ushort y)
    {
        X = x;
        Y = y;
    }

    public override InputEventKind EventKind => InputEventKind.MouseMove;

    /// <summary>Normalized 0-65535, virtual-display coordinate space.</summary>
    public ushort X { get; }

    /// <summary>Normalized 0-65535, virtual-display coordinate space.</summary>
    public ushort Y { get; }

    public override byte[] Encode()
    {
        var buffer = new byte[5];
        buffer[0] = (byte)EventKind;
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(1, 2), X);
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(3, 2), Y);
        return buffer;
    }
}

/// <summary>SPEC.md §4.2 0x02 Mouse button down (6 bytes).</summary>
public sealed class MouseButtonDownEvent : InputEvent
{
    public MouseButtonDownEvent(ushort x, ushort y, MouseButton button)
    {
        X = x;
        Y = y;
        Button = button;
    }

    public override InputEventKind EventKind => InputEventKind.MouseButtonDown;
    public ushort X { get; }
    public ushort Y { get; }
    public MouseButton Button { get; }

    public override byte[] Encode()
    {
        var buffer = new byte[6];
        buffer[0] = (byte)EventKind;
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(1, 2), X);
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(3, 2), Y);
        buffer[5] = (byte)Button;
        return buffer;
    }
}

/// <summary>SPEC.md §4.2 0x03 Mouse button up (6 bytes).</summary>
public sealed class MouseButtonUpEvent : InputEvent
{
    public MouseButtonUpEvent(ushort x, ushort y, MouseButton button)
    {
        X = x;
        Y = y;
        Button = button;
    }

    public override InputEventKind EventKind => InputEventKind.MouseButtonUp;
    public ushort X { get; }
    public ushort Y { get; }
    public MouseButton Button { get; }

    public override byte[] Encode()
    {
        var buffer = new byte[6];
        buffer[0] = (byte)EventKind;
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(1, 2), X);
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(3, 2), Y);
        buffer[5] = (byte)Button;
        return buffer;
    }
}

/// <summary>SPEC.md §4.2 0x04 Mouse wheel (9 bytes).</summary>
public sealed class MouseWheelEvent : InputEvent
{
    public MouseWheelEvent(ushort x, ushort y, short deltaX, short deltaY)
    {
        X = x;
        Y = y;
        DeltaX = deltaX;
        DeltaY = deltaY;
    }

    public override InputEventKind EventKind => InputEventKind.MouseWheel;
    public ushort X { get; }
    public ushort Y { get; }

    /// <summary>Signed; positive = right.</summary>
    public short DeltaX { get; }

    /// <summary>Signed; positive = up.</summary>
    public short DeltaY { get; }

    public override byte[] Encode()
    {
        var buffer = new byte[9];
        buffer[0] = (byte)EventKind;
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(1, 2), X);
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(3, 2), Y);
        BinaryPrimitives.WriteInt16BigEndian(buffer.AsSpan(5, 2), DeltaX);
        BinaryPrimitives.WriteInt16BigEndian(buffer.AsSpan(7, 2), DeltaY);
        return buffer;
    }
}

/// <summary>SPEC.md §4.2 0x05 Key down (3 bytes). <see cref="HidUsage"/> is a USB HID Usage ID, Usage Page 0x07.</summary>
public sealed class KeyDownEvent : InputEvent
{
    public KeyDownEvent(ushort hidUsage)
    {
        HidUsage = hidUsage;
    }

    public override InputEventKind EventKind => InputEventKind.KeyDown;
    public ushort HidUsage { get; }

    public override byte[] Encode()
    {
        var buffer = new byte[3];
        buffer[0] = (byte)EventKind;
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(1, 2), HidUsage);
        return buffer;
    }
}

/// <summary>SPEC.md §4.2 0x06 Key up (3 bytes). <see cref="HidUsage"/> is a USB HID Usage ID, Usage Page 0x07.</summary>
public sealed class KeyUpEvent : InputEvent
{
    public KeyUpEvent(ushort hidUsage)
    {
        HidUsage = hidUsage;
    }

    public override InputEventKind EventKind => InputEventKind.KeyUp;
    public ushort HidUsage { get; }

    public override byte[] Encode()
    {
        var buffer = new byte[3];
        buffer[0] = (byte)EventKind;
        BinaryPrimitives.WriteUInt16BigEndian(buffer.AsSpan(1, 2), HidUsage);
        return buffer;
    }
}

File: protocol/swift/ToInfinityProtocol/Sources/ToInfinityProtocol/InputEvent.swift
import Foundation

/// Mouse button identifier for `.mouseButtonDown`/`.mouseButtonUp` (SPEC.md §4.2).
/// Modeled as a raw-byte wrapper (not a closed `enum`) so an unrecognized
/// button value still round-trips rather than failing to decode, matching
/// the C# `MouseButton` enum's permissive byte cast semantics.
public struct MouseButton: RawRepresentable, Equatable {
    public let rawValue: UInt8

    public init(rawValue: UInt8) {
        self.rawValue = rawValue
    }

    public static let left = MouseButton(rawValue: 0)
    public static let right = MouseButton(rawValue: 1)
    public static let middle = MouseButton(rawValue: 2)
    public static let x1 = MouseButton(rawValue: 3)
    public static let x2 = MouseButton(rawValue: 4)
}

/// SPEC.md §4.2 input event kind tags.
public enum InputEventKind: UInt8 {
    case mouseMove = 0x01
    case mouseButtonDown = 0x02
    case mouseButtonUp = 0x03
    case mouseWheel = 0x04
    case keyDown = 0x05
    case keyUp = 0x06
}

/// Binary input events carried on the control channel with
/// `ControlFrameKind.inputEvent` (SPEC.md §4). Mouse coordinates are
/// normalized 0-65535 per axis (SPEC.md §4.1); key codes are USB HID Usage
/// IDs, Usage Page 0x07 (SPEC.md §4.2).
public enum InputEvent: Equatable {
    /// SPEC.md §4.2 0x01 (5 bytes). x/y normalized 0-65535, virtual-display coordinate space.
    case mouseMove(x: UInt16, y: UInt16)
    /// SPEC.md §4.2 0x02 (6 bytes).
    case mouseButtonDown(x: UInt16, y: UInt16, button: MouseButton)
    /// SPEC.md §4.2 0x03 (6 bytes).
    case mouseButtonUp(x: UInt16, y: UInt16, button: MouseButton)
    /// SPEC.md §4.2 0x04 (9 bytes). deltaX positive = right, deltaY positive = up.
    case mouseWheel(x: UInt16, y: UInt16, deltaX: Int16, deltaY: Int16)
    /// SPEC.md §4.2 0x05 (3 bytes). hidUsage is a USB HID Usage ID, Usage Page 0x07.
    case keyDown(hidUsage: UInt16)
    /// SPEC.md §4.2 0x06 (3 bytes). hidUsage is a USB HID Usage ID, Usage Page 0x07.
    case keyUp(hidUsage: UInt16)

    public var eventKind: InputEventKind {
        switch self {
        case .mouseMove: return .mouseMove
        case .mouseButtonDown: return .mouseButtonDown
        case .mouseButtonUp: return .mouseButtonUp
        case .mouseWheel: return .mouseWheel
        case .keyDown: return .keyDown
        case .keyUp: return .keyUp
        }
    }

    /// Encodes this event's payload, starting with the 1-byte event kind tag.
    public func encode() -> Data {
        var bytes: [UInt8]
        switch self {
        case let .mouseMove(x, y):
            bytes = [eventKind.rawValue]
            bytes.append(contentsOf: BigEndian.writeUInt16(x))
            bytes.append(contentsOf: BigEndian.writeUInt16(y))

        case let .mouseButtonDown(x, y, button):
            bytes = [eventKind.rawValue]
            bytes.append(contentsOf: BigEndian.writeUInt16(x))
            bytes.append(contentsOf: BigEndian.writeUInt16(y))
            bytes.append(button.rawValue)

        case let .mouseButtonUp(x, y, button):
            bytes = [eventKind.rawValue]
            bytes.append(contentsOf: BigEndian.writeUInt16(x))
            bytes.append(contentsOf: BigEndian.writeUInt16(y))
            bytes.append(button.rawValue)

        case let .mouseWheel(x, y, deltaX, deltaY):
            bytes = [eventKind.rawValue]
            bytes.append(contentsOf: BigEndian.writeUInt16(x))
            bytes.append(contentsOf: BigEndian.writeUInt16(y))
            bytes.append(contentsOf: BigEndian.writeInt16(deltaX))
            bytes.append(contentsOf: BigEndian.writeInt16(deltaY))

        case let .keyDown(hidUsage):
            bytes = [eventKind.rawValue]
            bytes.append(contentsOf: BigEndian.writeUInt16(hidUsage))

        case let .keyUp(hidUsage):
            bytes = [eventKind.rawValue]
            bytes.append(contentsOf: BigEndian.writeUInt16(hidUsage))
        }
        return Data(bytes)
    }

    /// Attempts to decode an input event payload (the bytes after the
    /// `ControlFrameKind.inputEvent` tag). Returns nil (never throws) for
    /// an unrecognized event kind or a payload too short for its kind, per
    /// SPEC.md §6.
    public static func decode(from payload: Data) -> InputEvent? {
        guard payload.count >= 1 else {
            return nil
        }

        let bytes = [UInt8](payload)
        guard let kind = InputEventKind(rawValue: bytes[0]) else {
            return nil
        }

        switch kind {
        case .mouseMove:
            guard bytes.count == 5 else { return nil }
            let x = BigEndian.readUInt16(bytes, at: 1)
            let y = BigEndian.readUInt16(bytes, at: 3)
            return .mouseMove(x: x, y: y)

        case .mouseButtonDown:
            guard bytes.count == 6 else { return nil }
            let x = BigEndian.readUInt16(bytes, at: 1)
            let y = BigEndian.readUInt16(bytes, at: 3)
            return .mouseButtonDown(x: x, y: y, button: MouseButton(rawValue: bytes[5]))

        case .mouseButtonUp:
            guard bytes.count == 6 else { return nil }
            let x = BigEndian.readUInt16(bytes, at: 1)
            let y = BigEndian.readUInt16(bytes, at: 3)
            return .mouseButtonUp(x: x, y: y, button: MouseButton(rawValue: bytes[5]))

        case .mouseWheel:
            guard bytes.count == 9 else { return nil }
            let x = BigEndian.readUInt16(bytes, at: 1)
            let y = BigEndian.readUInt16(bytes, at: 3)
            let deltaX = BigEndian.readInt16(bytes, at: 5)
            let deltaY = BigEndian.readInt16(bytes, at: 7)
            return .mouseWheel(x: x, y: y, deltaX: deltaX, deltaY: deltaY)

        case .keyDown:
            guard bytes.count == 3 else { return nil }
            return .keyDown(hidUsage: BigEndian.readUInt16(bytes, at: 1))

        case .keyUp:
            guard bytes.count == 3 else { return nil }
            return .keyUp(hidUsage: BigEndian.readUInt16(bytes, at: 1))
        }
    }
}
